### PR TITLE
feat: enhance `package.json` finder

### PR DESCRIPTION
**Pull request by reey**
implements #1284

For assets loaded from a subdirectory of `node_modules`, it will pick the first `package.json` that actually has `name` and `version` attributes.
Added a testcase for this, which verifies this functionality for the `luxon` and `libphonenumber-js` packages.
The `package.json` for `libphonenumber-js/max` [misses a version number](https://unpkg.com/browse/libphonenumber-js@1.11.3/max/package.json) while for `luxon` [the name is missing](https://unpkg.com/browse/luxon@3.4.4/src/package.json).

The snapshot has quite a lot of changes:
- For some reason the `purl` entries are no longer url encoded.
- The `@apollo/client/*` entries have been merged into a single `@apollo/client` entry
- The `@babel/runtime` was added as dependency

### CODE REVIEW

File: src/_helpers.ts [hunk 1]
```diff
@@ -18,36 +18,56 @@ Copyright (c) OWASP Foundation. All Rights Reserved.
...
-    if (nextPath === path) {
+    if (nextPath === path || isNodeModulesFolder(nextPath)) {
       return undefined
     }
     path = nextPath
   }
   return undefined
 }
 
+function isNodeModulesFolder (path: string): boolean {
```
Comment (jkowalleck) [inline src/_helpers.ts:57]: could you also add a test for yarn/pnp? 
```js
function isPnpRoot(path: string): boolean {
  return path.endsWith('.zip')
}
```

**Comment by jkowalleck** [comment]
Thanks for the implementation, @reey .


I really like the solution, but it needs to be safe and waterproof. 
The thing with the "enhancement" you plan on introducing is: it opens up for a whole world of false-positives and edge cases.
Let's take the time and make sure it meets the needed edge cases. :-) 

For additional tests, we need to have a 
- yarn2 setup with the edge case s in unplugged mode
- yarn2 setup with the edge cases in non-unplugged (cached) mode
- yarn2/pnp setup with the edge case s in unplugged mode
- yarn2/pnp setup with the edge cases in non-unplugged (cached) mode
- all edge cases with pnpm ... dont knwon what exists, there
- all edge cases with "external" linked-in paths ... have fun figguring out what edge cases might exists there ...

**Comment by reey** [comment]
@jkowalleck Thanks for the feedback.
I've done some testing with yarn, but it should not be any different as the (folder) structure still has a node_modules directory:
`/home/<user>/.yarn/berry/cache/libphonenumber-js-npm-1.11.3-8309751739-10c0.zip/node_modules/libphonenumber-js/build/legacy`.

The overall idea was if there would be no `node_modules` for some package manager, to not change the behavior from how it currently works.
In case changes are required for a specific package manager, this could still be done.

Regarding the URL encoded purls it seems that the behavior is different for windows vs linux? is this intended?

File: src/_helpers.ts (hunk 1 above)
Comment (jkowalleck) [reply src/_helpers.ts:57]: not needed, as stated here: https://github.com/CycloneDX/cyclonedx-webpack-plugin/pull/1286#issuecomment-2168748611

File: tests/integration/setup.js [hunk 2]
```diff
@@ -56,4 +61,19 @@ const path = require('path');
       console.error(done)
     }
   }
+
+  for (const DIR of REQUIRES_YARN_INSTALL) {
+    console.log('>>> setup with yarn:', DIR)
+    done = spawnSync(
+      'yarn', ['install', '--immutable'], {
```
Comment (jkowalleck) [inline tests/integration/setup.js:68]: yarn lock file is missing, so `--immutable` might have no effect.
@reey, please ad a yarn lock to `tests/integration/improvement-issue-1284-yarn/`?

File: tests/integration/improvement-issue-1284-yarn/.gitignore [hunk 3]
```diff
@@ -0,0 +1,9 @@
+*
+!/.gitignore
+!/.gitattributes
+!/README.md
+!/package.json
+!/package-lock.json
```
Comment (jkowalleck) [inline tests/integration/improvement-issue-1284-yarn/.gitignore:6]: i guess this should be the `yarn` lock instead.

**Review by jkowalleck — CHANGES_REQUESTED** [review]

**Comment by jkowalleck** [comment]
@reey, your implementation looks solid, your test beds are great!
a thing that might be missing: 
add a lock file to the yarn test bed, and add a `yarn set version ...` before the `yarn install` happens for the test beds.

File: tsconfig.json [hunk 4]
```diff
@@ -99,7 +99,7 @@
 
     /* Completeness */
     "skipDefaultLibCheck": false,                      /* Skip type checking .d.ts files that are included with TypeScript. */
-    "skipLibCheck": false                                 /* Skip type checking all .d.ts files. */
+    "skipLibCheck": true                                 /* Skip type checking all .d.ts files. */
```
Comment (jkowalleck) [inline tsconfig.json:102]: why?

**Review by jkowalleck — APPROVED** [review]
all looks good,
i will fix the `yarn` situation in another branch first,
and will merge your feature afterward.

File: tsconfig.json (hunk 4 above)
Comment (reey) [reply tsconfig.json]: Had some issues there on my windows machine..
Reverted that

**Comment by jkowalleck** [comment]
Thanks for the contribution, 
it was released via https://github.com/CycloneDX/cyclonedx-webpack-plugin/releases/tag/v3.12.0

File: tsconfig.json (hunk 4 above)
Comment (reey) [reply tsconfig.json]: FYI, this was the issue on windows:
```
npm run build

> @cyclonedx/webpack-plugin@3.12.0 build
> run-p --aggregate-output -l build:*   

[build:node] 
[build:node] > @cyclonedx/webpack-plugin@3.12.0 prebuild:node
[build:node] > node -r fs -e 'fs.rmSync("dist",{recursive:true,force:true})'
[build:node] 
[build:node] 
[build:node] > @cyclonedx/webpack-plugin@3.12.0 build:node
[build:node] > tsc -b ./tsconfig.json
[build:node]
[build:node] node_modules/@types/node/globals.d.ts(72,13): error TS2403: Subsequent variable declarations must have the same type.  Variable 'AbortSignal' must be of type '{ new (): AbortSignal; prototype: AbortSignal; abort(reason?: any): AbortSignal; timeout(milliseconds: number): AbortSignal; }', but here has type '{ new (): AbortSignal; prototype: AbortSignal; }'.
ERROR: "build:node" exited with 1.
```

Comment (jkowalleck) [reply tsconfig.json]: ah, thanks, will add a windows CI/CT to have this coverred.
